### PR TITLE
SI-9315 Desugar string concat to java.lang.StringBuilder ...

### PR DESCRIPTION
**Pull request by soc**
... instead of scala.collection.mutable.StringBuilder to benefit from
JVM optimizations.

### CODE REVIEW

**Comment by soc** [comment]
Thanks a lot, @Blaisorblade, the pasto was the remaining issue! :-)
Do you want to review it?

**Comment by Blaisorblade** [comment]
I'm happy I was of help!
The commit looks good to me at first sight, but I don't feel qualified for a real review. I'll add some comments anyway.

**Comment by Blaisorblade** [comment]
That's all I had to say. Kudos, and good luck with the next reviewer!

**Comment by soc** [comment]
`StringBuilderClass` is still used by the old GenASM backend. Given that the tests (afaik) only run against the new backend, I felt that the best way of action was to not touch the old backend to prevent introducing issues in less well-tested places of the code.

**Comment by Blaisorblade** [comment]
> Given that the tests (afaik) only run against the new backend, I felt that the best way of action was to not touch the old backend to prevent introducing issues in less well-tested places of the code.

:+1:

**Comment by SethTisue** [comment]
review by @lrytz?

**Comment by lrytz** [comment]
sure. sorry i was a bit off the pr queue lately..

**Comment by retronym** [comment]
While we're looking at this part of the compiler, I'd like to consider making `def foo(c: CharSequence) = " " + c` dispatch to the suitable overloaded `append` to avoid the `toString` call.

This would be a deviation from Java, but it would be handy for things like `Name` (assuming we made it implement `CharSequence`) to avoid temporary strings.

**Comment by soc** [comment]
@retronym Not sure I understand this completely. Can you describe how the calls should look under the new scheme (vs. the old)?

**Comment by soc** [comment]
Mhhh, do you mean https://github.com/soc/scala/blob/SI-9315/src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala#L223, where it always picks `Object` for reference types?

**Comment by soc** [comment]
Mhhh, I guess we should do the same for arguments of type `StringBuilder`...

**Comment by retronym** [comment]
Yes. The parameter type corresponds to signature of the fake method `+` we install on the type `java.lang.String`. We could add a second overload to `+`, although that technically is source incompatible. Or, we could perform the dispatch in the backend, based on the static type of the argument.

**Comment by retronym** [comment]
This change of course would be a behavioural difference for ill-behaved `CharSequence` implementations. Java string concatenation is stuck with the status quo for compatibiity reasons. I'm not sure the gain is so important that we'll take the risk of the change ourselves, but I just wanted to raise awareness of it and invite discussion.

**Comment by Blaisorblade** [comment]
> This change of course would be a behavioural difference for ill-behaved CharSequence implementations.

I'd argue that, even for 2.12, the behavior should be equivalent for `CharSequence` implementations that respect the `CharSequence` specification. That includes even the odd `CharBuffer`, which is called out by the `CharSequence` docs, so we're fine? Or are there notable exceptions?

**Comment by lrytz** [comment]
I think you could quite easily do it in the backend (not tested!):

``` diff
diff --git a/src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala b/src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala
index 535e1a8..99efb0d 100644
--- a/src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala
+++ b/src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala
@@ -219,9 +219,11 @@ abstract class BCodeIdiomatic extends SubComponent {
      */
     final def genStringConcat(el: BType, pos: Position): Unit = {

-      val jtype =
-        if (el.isArray || el.isClass) ObjectReference
-        else el
+      val jtype = el match {
+        case c: ClassBType if c.isSubtypeOf(CharSequenceReference) => CharSequenceReference
+        case t if t.isArray || t.isClass => ObjectReference
+        case t => t
+      }

       val bt = MethodBType(List(jtype), StringBuilderReference)

diff --git a/src/compiler/scala/tools/nsc/backend/jvm/CoreBTypes.scala b/src/compiler/scala/tools/nsc/backend/jvm/CoreBTypes.scala
index 00ca096..c0339de 100644
--- a/src/compiler/scala/tools/nsc/backend/jvm/CoreBTypes.scala
+++ b/src/compiler/scala/tools/nsc/backend/jvm/CoreBTypes.scala
@@ -108,6 +108,7 @@ class CoreBTypes[BTFS <: BTypesFromSymbols[_ <: Global]](val bTypes: BTFS) {

   lazy val StringReference             : ClassBType = classBTypeFromSymbol(StringClass)
   lazy val StringBuilderReference      : ClassBType = classBTypeFromSymbol(StringBuilderClass)
+  lazy val CharSequenceReference       : ClassBType = classBTypeFromSymbol(requiredClass[CharSequence])
   lazy val ThrowableReference          : ClassBType = classBTypeFromSymbol(ThrowableClass)
   lazy val jlCloneableReference        : ClassBType = classBTypeFromSymbol(JavaCloneableClass)        // java/lang/Cloneable
   lazy val jlNPEReference              : ClassBType = classBTypeFromSymbol(NullPointerExceptionClass) // java/lang/NullPointerException
@@ -254,6 +255,7 @@ final class CoreBTypesProxy[BTFS <: BTypesFromSymbols[_ <: Global]](val bTypes:

   def StringReference             : ClassBType = _coreBTypes.StringReference
   def StringBuilderReference      : ClassBType = _coreBTypes.StringBuilderReference
+  def CharSequenceReference       : ClassBType = _coreBTypes.CharSequenceReference
   def ThrowableReference          : ClassBType = _coreBTypes.ThrowableReference
   def jlCloneableReference        : ClassBType = _coreBTypes.jlCloneableReference
   def jlNPEReference              : ClassBType = _coreBTypes.jlNPEReference
```

**Comment by soc** [comment]
I also added support for the `StringBuffer` overload, but stopped short of the `char[]` one, I hope this is the right balance...

**Comment by lrytz** [comment]
Using the overload for `char[]` would be incorrect AFAICT: `toString` on a char is something like `[C@74a14482`, so that's the string that should be added when using a char array in a string concatenation. In other words, using the overload would change the current behavior:

```
scala> val cs = "hi".toArray
cs: Array[Char] = Array(h, i)

scala> "" + cs
res0: String = [C@6572421
```

**Comment by lrytz** [comment]
It would be great to add a test for the overloading change. The easiest would be a unit test in `test/junit/scala/tools/nsc/backend/jvm`, you can use the tools in [CodeGenTools.scala](https://github.com/scala/scala/blob/2.11.x/test/junit/scala/tools/nsc/backend/jvm/CodeGenTools.scala), see for example [MethodLevelOpts.scala](https://github.com/scala/scala/blob/2.11.x/test/junit/scala/tools/nsc/backend/jvm/opt/MethodLevelOpts.scala)

**Comment by soc** [comment]
@lrytz Agree, that was my reasoning for leaving out `char[]`.
Thanks for the hints regarding testing, extremely useful!

**Comment by soc** [comment]
Is there anything needed to make junit pick the new tests up?

I wrote a test class, added all the annotations, but the class is never tested.

**Comment by lrytz** [comment]
it should happen by itself - maybe you missed some annotation? feel free to post the diff. btw, i very much recommend to write and run tests in intellij.

**Comment by soc** [comment]
@lrytz Ooops, sorry, forgot to push! Here are the tests: https://github.com/soc/scala/commit/2e05ec4bd81b0dbbf8664cb079128e2bc618e94c#diff-7506aaabf8fa132381745c90a77c1606R23

JUnit doesn't seem to pick up the class:

```
    ...
    [junit] Running scala.tools.nsc.backend.jvm.BTypesTest
    [junit] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.196 sec
    [junit] Running scala.tools.nsc.backend.jvm.DirectCompileTest
    [junit] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.341 sec
    [junit] Running scala.tools.nsc.backend.jvm.IndyLambdaTest
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.972 sec
    [junit] Running scala.tools.nsc.backend.jvm.analysis.NullnessAnalyzerTest
    [junit] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.972 sec
    [junit] Running scala.tools.nsc.backend.jvm.analysis.ProdConsAnalyzerTest
    [junit] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.083 sec
    [junit] Running scala.tools.nsc.backend.jvm.opt.BTypesFromClassfileTest
    ...
```

**Comment by lrytz** [comment]
no idea why it's not picked up.. i can run the test in intellij.

**Comment by lrytz** [comment]
@soc the answer is here: https://github.com/scala/scala/blob/2.11.x/build.xml#L1562. You need to name the class `StringConcatTest`.

**Comment by soc** [comment]
@lrytz Arrrgh, thanks a lot! :-)

**Comment by soc** [comment]
Ok, this is weird: `def result = "abc" + 1`

```
  public java.lang.String result();
    Code:
      stack=2, locals=1, args_size=1
         0: new           #17                 // class java/lang/StringBuilder
         3: dup
         4: invokespecial #21                 // Method java/lang/StringBuilder."<init>":()V
         7: ldc           #23                 // String abc
         9: invokevirtual #27                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
        12: iconst_1
        13: invokestatic  #33                 // Method scala/runtime/BoxesRunTime.boxToInteger:(I)Ljava/lang/Integer;
        16: invokevirtual #36                 // Method java/lang/StringBuilder.append:(Ljava/lang/Object;)Ljava/lang/StringBuilder;
        19: invokevirtual #39                 // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
        22: areturn
```

I don't think it is intentional that we box the primitives and use `append(Object)`, right?

**Comment by lrytz** [comment]
That's why it's so great that finally somebody is writing tests for this thing :-)

Boxing is added in erasure.

```
$ scala -Xprint:erasure

scala> def result = "abc" + 1
[[syntax trees at end of                   erasure]] // <console>
package $line3 {
  object $read extends Object {
    def <init>(): $line3.$read.type = {
      $read.super.<init>();
      ()
    };
    object $iw extends Object {
      def <init>(): type = {
        $iw.super.<init>();
        ()
      };
      object $iw extends Object {
        def <init>(): type = {
          $iw.super.<init>();
          ()
        };
        def result(): String = "abc".+(scala.Int.box(1))
      }
    }
  }
}
```

It must be because there's only a single `+` method "added" to the string class: https://github.com/scala/scala/blob/2.11.x/src/reflect/scala/reflect/internal/Definitions.scala#L1061. We should add overloads for the primitives.

**Comment by lrytz** [comment]
here's a starting point: https://github.com/lrytz/scala/commit/e90fb97bffec2109b32efb062263bedaf1cca8a0

**Comment by retronym** [comment]
On a similar note, we box on:

``` scala
class C { def syncHash: Int = synchronized(hashCode) }
```

Overloading `String#+` or `Object#synchronized` would certainly be a possible solution.

Another might be to treat them as having a `@PolymorphicSignature`, which we now support for `MethodHandle#{invoke,invokeExact}`

**Comment by lrytz** [comment]
> `@PolymorphicSignature`

I wasn't aware of this. I think it's the preferable solution.

**Comment by soc** [comment]
@retronym Wouldn't this mean that we would keep entering methods into the `String` just to discard them later, except for the 12 existing ones? (8 primitives + Object + String + CharSequqnce + StringBuffer)

**Comment by lrytz** [comment]
@soc i just looked at the current implementation of `@PolymorphicSignature` support (https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3303). So there's a new symbol created for every callsite of a signature-polymorphic method. Given the frequency of string concat, I also think that solution is too heavy. So I'd proceed with adding the overloads for `+` to the `String` class in definitions.

**Comment by soc** [comment]
@lrytz Yes, that was my concern. Will do manual overloads! Thanks!

**Comment by soc** [comment]
I'm starting to wonder whether it would have been better to avoid doing any optimizations at all, and optimize with the BCode optimizers...

**Comment by soc** [comment]
With the WIP added, I'm seeing various test failures, like:

```
# starting 3 tests in run
!! 1 - run/inline-ex-handlers.scala              [compilation failed]
!! 2 - run/t6240-universe-code-gen.scala         [non-zero exit code]
!! 3 - run/reflection-magicsymbols-invoke.scala  [output differs]
# 0/3 passed, 3 failed in run

# starting 1 test in jvm
!! 1 - jvm/try-type-tests.scala                  [compilation failed]
# 0/1 passed, 1 failed in jvm
```

Example `jvm/try-type-tests.scala`:

```
  def testFoldFailure(): Unit = {
    val t = Failure(new Exception("foo"))
    val res = t.fold("Throws " + _, "Returns " + _)
    assert(res == "Throws java.lang.Exception: foo")
  }
```

```
try-type-tests.scala:139: error: ambiguous reference to overloaded definition,
both method + in class String of type (x$1: Boolean)String
and  method + in class String of type (x$1: Byte)String
match argument types (Nothing)
    val res = t.fold("Throws " + _, "Returns " + _)
                                               ^
```

```
  def testFoldFailureFailure(): Unit = {
    val t = Failure(new Exception("foo"))
    val res = try {
      t.fold(_ => throw new Exception("bar"), "Returns " + _)
    } catch {
      case e: Throwable => "Throws " + e
    }
    assert(res == "Throws java.lang.Exception: bar")
  }
```

```
try-type-tests.scala:152: error: ambiguous reference to overloaded definition,
both method + in class String of type (x$1: Boolean)String
and  method + in class String of type (x$1: Byte)String
match argument types (Nothing)
      t.fold(_ => throw new Exception("bar"), "Returns " + _)
                                                         ^
```

**Comment by lrytz** [comment]
that's too bad.. how about adding an overload for `Nothing`?

```
scala> object T { def f(x: Int) = 0; def f(x: Object) = 1 }
defined object T

scala> Nil.map(T.f(_))
<console>:12: error: ambiguous reference to overloaded definition,
both method f in object T of type (x: Object)Int
and  method f in object T of type (x: Int)Int
match argument types (Nothing)
       Nil.map(T.f(_))
                 ^

scala> object TT { def f(x: Int) = 0; def f(x: Object) = 1; def f(n: Nothing) = 2 }
defined object TT

scala> Nil.map(TT.f(_))
res8: List[Int] = List()
```

**Comment by lrytz** [comment]
I'll think about your idea of handling it in the optimizer, maybe that's easier / cleaner!

**Comment by soc** [comment]
What I would propose is adopting the incremental improvement in `StringBuilder` for now, and deferring the additional optimizations for `String.valueOf` until I have a better overview of the solution space.

The most pressing question here is probably what happens to ICode/GenASM in 2.12, because depending on the approach testing/dragging along all the old code will be a substantial cost.

Afaik, ICode/GenASM have not been updated for Java 8, so producing code targeting Java 8 would be incompatible with code emitted by GenBCode, and emitting code targeting Java 6 won't work either, right?

**Comment by soc** [comment]
Ok, I dropped all but the first commit. I'll revisit this after the GenASM/ICode cleanup, I'll probably try and take the GenBCode optimizer route.

**Comment by lrytz** [comment]
OK, I agree we can postpone specialization. However, could you file a ticket about primitive boxing in string concatenation, add a comment [here](https://github.com/scala/scala/pull/4737/files#diff-98a5969a8ecd1feba56b73725add24fcR226) and also add a test for the status quo (test that primitives are boxed)?

**Comment by soc** [comment]
@DarkDimius Thanks!
@lrytz I recommend waiting until I added some tests before merging the backport!

**Comment by lrytz** [comment]
@soc this needs a rebase. Once the TODOs in https://github.com/scala/scala/pull/4737#issuecomment-141972190 are addressed I think we can get this in!

**Comment by lrytz** [comment]
@soc do you have time to push this one forward? it would be a nice addition to 2.12.0-M4 (probably mid-february).

**Comment by soc** [comment]
@lrytz Rebased, addressed comments, added test.

File: src/compiler/scala/tools/nsc/backend/jvm/CoreBTypes.scala [hunk 1]
```diff
@@ -95,7 +95,10 @@ class CoreBTypes[BTFS <: BTypesFromSymbols[_ <: Global]](val bTypes: BTFS) {
   lazy val ObjectRef                 : ClassBType = classBTypeFromSymbol(ObjectClass)
   lazy val StringRef                 : ClassBType = classBTypeFromSymbol(StringClass)
   lazy val PredefRef                 : ClassBType = classBTypeFromSymbol(PredefModule.moduleClass)
-  lazy val jlStringBuilderRef        : ClassBType = classBTypeFromSymbol(StringBuilderClass)
+  lazy val StringBuilderRef          : ClassBType = classBTypeFromSymbol(StringBuilderClass)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/CoreBTypes.scala]: `StringBuilderRef` seems to be not used

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala [hunk 2]
```diff
@@ -40,7 +40,7 @@ abstract class BCodeIdiomatic extends SubComponent {
     if (emitStackMapFrame) asm.ClassWriter.COMPUTE_FRAMES else 0
   )
 
-  val StringBuilderClassName = "scala/collection/mutable/StringBuilder"
+  val JavaStringBuilderClassName = "java/lang/StringBuilder"
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala]: could be `def JavaStringBuilderClassName = jlStringBuilderRef.internalName` (or `lazy val`)

File: src/compiler/scala/tools/nsc/backend/jvm/opt/BoxUnbox.scala [hunk 3]
```diff
@@ -21,7 +21,7 @@ class BoxUnbox[BT <: BTypes](val btypes: BT) {
   import backendUtils._
 
   /**
-   * Eliminate box-unbox paris within `method`. Such appear commonly after closure elimination:
+   * Eliminate box-unbox parts within `method`. Such appear commonly after closure elimination:
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/BoxUnbox.scala]: should be `pairs` :smile:

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 4]
```diff
@@ -428,13 +428,16 @@ trait Definitions extends api.StandardDefinitions {
...
-    lazy val ListClass          = requiredClass[scala.collection.immutable.List[_]]
-    lazy val SeqClass           = requiredClass[scala.collection.Seq[_]]
-    lazy val StringBuilderClass = requiredClass[scala.collection.mutable.StringBuilder]
-    lazy val TraversableClass   = requiredClass[scala.collection.Traversable[_]]
+    lazy val ConsClass              = requiredClass[scala.collection.immutable.::[_]]
+    lazy val IteratorClass          = requiredClass[scala.collection.Iterator[_]]
+    lazy val IterableClass          = requiredClass[scala.collection.Iterable[_]]
+    lazy val ListClass              = requiredClass[scala.collection.immutable.List[_]]
+    lazy val SeqClass               = requiredClass[scala.collection.Seq[_]]
+    lazy val StringBuilderClass     = requiredClass[scala.collection.mutable.StringBuilder]
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/Definitions.scala]: can this (`StringBuilderClass`) be removed now?

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala (hunk 2 above)
Comment (soc) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala]: I made it a `val` considering that it's highly unlikely that we don't need it, and probably need it more than once, if that's fine for you?

**Comment by soc** [comment]
@lrytz Done!

**Comment by lrytz** [comment]
LGTM, thanks for another iteration!

**Pull request merged**